# Bài tập 5

Bùi Văn Hợp - 1712046

## Câu 1 (1 điểm)

Ta có:

$$\mathbb{E}_{\mathcal{D}}[E_{in}(\mathbf{w}_{lin})] = \sigma^2\left(1 - \frac{d+1}{N}\right) ~~~~ (1)$$

$\sigma = 0.1, d = 8$

Từ (1) ta có thể rút gọn biểu thức và đưa ra giới hạn cho N là:

$$N > \frac{d + 1}{1 - \frac{0.008}{\sigma^2}}$$

Thay giá trị $\sigma, d$, ta có:

$$N > 45$$

**Với các đáp án trong bài, mình lựa chọn đáp án nhỏ nhất và thỏa mãn là 100, c.**

YOUR ANSWER HERE c

## Câu 2 (1 điểm)

Với hypothesis được đưa ra bởi bài 2, đó là 1 hyperbol có dạng như sau:

![hyperbola.gif](hyperbola.gif)

Với 2 trục x1, x2 ta có dạng của hyperbol là:
$$\frac{x_1^2}{a} - \frac{x_2^2}{b} = 1$$

Hay ta có thể biến đổi về dạng:

$$ax_1^2 - bx_2^2 - ab = 0 ~~~~ (1)$$

Tuy nhiên với miền giá trị, biểu thức $ax_1^2 - bx_2^2 - ab > 0$ với miền giá trị sau khi giải ra là:

$$x_1 > \sqrt{a}$$ hoặc $$x_1 < \sqrt{a}$$ với $x_2$ trong khoảng : $$\frac{-\sqrt{b}\sqrt{x_1^2-a}}{\sqrt{a}} < y < \frac{\sqrt{b}\sqrt{x_1^2-a}}{\sqrt{a}}$$

Hay qua hình sau với phần màu xám là phần dương của hypothesis:

![rangepositive.png](rangepositive.png)

Vậy nên để chính xác với miền dữ liệu +1 và -1 trong bài 2, ta cần đảo dấu của hypothesis của (1) lại là:

$$h(x) = -(ax_1^2 - bx_2^2 - ab) = -ax_1^2 + bx_2^2 + ab $$

Từ $h(x)$ ta có thể biết được $w_1 < 0$ và $w_2 > 0$. Mình sẽ lựa chọn đáp án d.

YOUR ANSWER HERE d

## Câu 3 (1 điểm)

Với phép biến đổi thành các dạng đa thức bậc 4:

$$\Phi_4 = x → (1, x_1, x_2, x_1^2, x_1x_2, x_2^2, x_1^3,x_1^2x_2,x_1x_2^2, x_2^3, x_1^4, x_1^3x_2, x_1^2x_2^2, x_1x_2^3, x_2^4)$$

Ta có số chiều của $\Phi_4$ là 14 (không kể $z_0$). **Nên $d_{VC}$ của linear model với phép biến đổi phi tuyến này là 15. Đáp án c.**

YOUR ANSWER HERE c

## Câu 4 (1 điểm)

Ta có hàm lỗi theo u và v là:

$$E(u, v) = (ue^{v} - 2ve^{-u})^2$$

Từ đây ta tính được các đạo hàm riêng theo u, v là:

$$\frac{\partial E}{\partial u} = 2(e^v + 2ve^{-u})(ue^v - 2ve^{-u})$$

$$\frac{\partial E}{\partial v} = 2(ue^v - 2e^{-u})(ue^v - 2ve^{-u})$$

Đáp án mình lựa chọn cho câu 4 là e.

YOUR ANSWER HERE e

## Câu 5 (1 điểm)

In [1]:
import numpy as np

In [2]:
def get_error(u, v):
    return (u * np.exp(v) - 2 * v * np.exp(-u))**2

def get_partial_derriviate_u(u, v):
    return 2 * (np.exp(v) + 2 * v * np.exp(-u)) * (u * np.exp(v) - 2 * v * np.exp(-u))

def get_partial_derriviate_v(u, v):
    return 2 * (u * np.exp(v) - 2 * np.exp(-u)) * (u * np.exp(v) - 2 * v * np.exp(-u))

In [3]:
u, v = 1, 1
lr = 0.1
iteration = 0

while get_error(u,v) > 1e-14:
    du = get_partial_derriviate_u(u, v)
    dv = get_partial_derriviate_v(u, v)
    u -= lr * du
    v -= lr * dv
    iteration += 1
print("Iteration : ", iteration)
print("({}, {}) when converged".format(u, v))

Iteration :  10
(0.04473629039778207, 0.023958714099141746) when converged


Như kết quả chạy ở trên, sau 10 vòng lặp sử dụng Gradient Descent thì độ lỗi giảm xuống dưới 1e-14. **Vậy đáp án mình chọn là d, 10.**

YOUR ANSWER HERE d

## Câu 6 (1 điểm)

(u,v) sau khi chạy chương trình là (0.04473629039778207, 0.023958714099141746).

**Vậy mình lựa chọn đáp án gần với kết quả nhất là e, (0.045, 0.024).**

YOUR ANSWER HERE e

## Câu 7 (1 điểm)

In [4]:
u, v = 1, 1
lr = 0.1

for _ in range(15):
    u -= lr * get_partial_derriviate_u(u, v)
    v -= lr * get_partial_derriviate_v(u, v)
print('Error after 15 iterations with coordinate descent : ', get_error(u, v))

Error after 15 iterations with coordinate descent :  0.13981379199615315


Sau khi áp dụng "coordianate descent" (không phải gradient descent vì chúng ta cập nhật lần lượt từng biến và thay đổi đạo hàm cho biến tiếp theo) thì ta thấy độ lỗi sau 15 lần lặp là 0.13981379199615315. **Nên kết quả mình chọn gần với nó nhất là 0.1, a.**

Một giải thích cho vì sao độ lỗi của cách descent này là chúng ta cập nhật các biến liên tục trong quá trình descent làm cho việc tính vector đạo hàm không còn chính xác cho các biến cũ cập nhất trước đó, dẫn tới đây không còn cách tối ưu tốt cho hàm số.

YOUR ANSWER HERE a

## Câu 8 (1 điểm)

In [5]:
def generate_target_w():
    """
    Generates target_w (the vector of parameters of f) 
    from two random, uniformly distributed points in [-1, 1] x [-1, 1].
    
    Returns
    -------
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    """
    # Generate two points from a uniform distribution over [-1, 1]x[-1, 1]
    p1 = np.random.uniform(-1, 1, 2)
    p2 = np.random.uniform(-1, 1, 2)
    # Compute the target W from these two points
    target_w = np.array([p1[1]*p2[0] - p1[0]*p2[1], p2[1] - p1[1], p1[0] - p2[0]]).reshape((-1, 1))
    
    return target_w

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def generate_data(N, target_w):
    """
    Generates a data set by generating random inputs and then using target_w to generate the 
    corresponding outputs.
    
    Parameters
    ----------
    N : int
        The number of examples.
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    
    Returns
    -------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.        
    """
    bad_data = True # `bad_data = True` means: data contain points on the target line 
                    # (this rarely happens, but just to be careful)
                    # -> y's of these points = 0 (with np.sign); 
                    #    we don't want this (y's of data must be -1 or 1)
                    # -> re-generate data until `bad_data = False`
    
    while bad_data == True:
        X = np.random.uniform(-1, 1, (N, 2))
        X = np.hstack((np.ones((N, 1)), X)) # Add 'ones' column
        z = sigmoid(np.dot(X, target_w))
        Y = np.zeros_like(z)
        Y[np.where(z > 0.5)] = 1
        Y[np.where(z <= 0.5)] = -1
        if (0 not in Y): # Good data
            bad_data = False
    
    return X, Y

In [6]:
def get_derrivation_w(x, y, w): #get derrivation respect to w
    dw = (np.exp(-y * np.dot(x, w)) * (-y * x.T)) / (1 + np.exp(-y * np.dot(x, w)))
    dw = np.expand_dims(dw, 1)
    return dw
def get_error_all(X, Y, w): #get error for all data X with label Y
    N = X.shape[0]
    E = np.sum(np.log(1 + np.exp(-Y * np.dot(X, w)))) / N
    return E

def run_LogisticRegression(X, Y, lr = 0.01, show_epoch = True):
    w = np.zeros((X.shape[1], 1))
    n_epochs = 0
    
    while True:
        np.random.seed(n_epochs)
        prev_w = w.copy()
        for i in np.random.permutation(X.shape[0]):
            x = X[i]
            y = Y[i]
            dw = get_derrivation_w(x, y, w)
            w -= lr * dw
        if show_epoch:
            print("Epoch ", n_epochs)
            print("Error after a epoch: ", get_error_all(X, Y, w))
        n_epochs += 1
        if (np.linalg.norm(prev_w - w) < 0.01):
            break
    return w, n_epochs

In [7]:
def main_LogisticRegression(N = 100, show_step = True, show_epoch = True):
    num_runs = 100
    avg_test_err = 0.0
    avg_epoch = 0.0
    lr = 0.01
    for r in range(num_runs):
        target_w = generate_target_w()
        X, Y = generate_data(N, target_w)
        
        w, n_epochs = run_LogisticRegression(X, Y, lr, show_epoch)
        X_test, Y_test = generate_data(10000, target_w)
        test_err = get_error_all(X_test, Y_test, w)
        if show_step:
            print('Test ', r)
            print('Number of epoch : ', n_epochs)
            print('E_out : ',test_err)
            print('-------------------')
        avg_test_err += test_err
        avg_epoch += n_epochs
    avg_test_err /= num_runs
    avg_epoch /= num_runs
    print('Average E_out', avg_test_err)
    print('Average number of epoch : ', avg_epoch)

In [8]:
main_LogisticRegression(100, True, False)

Test  0
Number of epoch :  317
E_out :  0.10948001796696273
-------------------
Test  1
Number of epoch :  374
E_out :  0.10753723419257771
-------------------
Test  2
Number of epoch :  379
E_out :  0.08818525148010697
-------------------
Test  3
Number of epoch :  304
E_out :  0.10679861325300508
-------------------
Test  4
Number of epoch :  267
E_out :  0.10764284071976277
-------------------
Test  5
Number of epoch :  262
E_out :  0.10890879920895272
-------------------
Test  6
Number of epoch :  359
E_out :  0.10361895442422335
-------------------
Test  7
Number of epoch :  372
E_out :  0.09936144748401786
-------------------
Test  8
Number of epoch :  330
E_out :  0.10218746356632806
-------------------
Test  9
Number of epoch :  373
E_out :  0.08761114451086259
-------------------
Test  10
Number of epoch :  334
E_out :  0.10905531131167553
-------------------
Test  11
Number of epoch :  350
E_out :  0.10203922726634547
-------------------
Test  12
Number of epoch :  383
E_out 

**Từ kết quả chạy trên, ta có trung bình của E_out xấp xỉ trong khoảng gần với 0.1. Nên mình sẽ lựa chọn đáp án là d, 0.100**

YOUR ANSWER HERE d

## Câu 9 (1 điểm)

Cũng từ Kết quả chạy, ta có trung bình số lượng các epoch để hoàn thành 1 lần chạy learning model Linear Regression cho bài toán này khoảng 350-355. **Nên mình sẽ lựa chọn đáp án a, 350 cho câu 9**

YOUR ANSWER HERE a

## Câu 10 (1 điểm)

Thuật toán PLA có thể được thực thi bằng SGD, tuy nhiên chúng ta cần 1 hàm lỗi chính xác để áp dụng SGD.

Với thuật toán PLA, ta có:

Với mỗi cặp $(x_n, y_n)$ nếu $sign(w^Tx_n) \neq y_n$ thì ta cập nhật lại $w = w + y_nx_n$

Ta có thể viết lại PLA:

$$sign(w^Tx_n) \neq y_n, w = w + y_nx_n$$

$$sign(w^Tx_n) = y_n, w = w + 0$$

Ta nhận thấy công thức trên của PLA có dạng các giá trị tùy theo từng điều kiện, nên ta có thể đoán được công thức cho $e_n(w)$ để áp dụng SGD cho PLA là đáp án e, có dạng:

$$e_n(w) = -\min(0,y_nw^Tx_n)$$

Từ đây ta có thể sử dụng SGD cho hàm lỗi với mỗi bước cập nhật như sau:

Từ công thức SGD, ta có: 
$$w = w - \eta \triangledown_w e_n(w)$$

$$w = w - \eta (-\min(0,y_nx_n))$$

Giả sử ở đây ta có thể đạo hàm được cho công thức $e_n(w) = -\min(0,y_nw^Tx_n)$ trên từng giá trị là 0 hoặc $y_nw^Tx_n$.

Ta đặt $\eta = 1$, công thức cập nhật w trở thành: 

$$w = w + \min(0,y_nx_n))$$

Ở đây ta sẽ phân tích thêm về công thức độ lỗi $e_n(w) = -\min(0,y_nw^Tx_n)$, 

Trường hợp 1: $sign(w^Tx_n) \neq y_n$, nghĩa là giá trị $w^Tx_n$ khác dấu với $y_n$ hay $y_nw^Tx_n < 0$, nên giá trị của công thức trên sẽ nhận là $y_nw^Tx_n$ và đạo hàm theo w sẽ là $y_nx_n$. Từ đây công thức cập nhật w sẽ là $w = w + y_nx_n ~~~~(1)$

Trường hợp 2: $sign(w^Tx_n) = y_n$, $y_nw^Tx_n > 0$ nên công thức độ lỗi sẽ nhận giá trị 0, ta cập nhật lại w là $w = w + 0 ~~~~(2)$

Từ (1) và (2) khi so sánh với công thức PLA qua 2 trường hợp ta sẽ thấy sự tương đồng. Nên mình sẽ lựa chọn hàm lỗi $e_n(w) = -\min(0,y_nw^Tx_n)$ là đáp án e.

YOUR ANSWER HERE e